src/PingData.h

```diff
@@ -9,12 +9,29 @@
 #ifndef SQUID_PINGDATA_H
 #define SQUID_PINGDATA_H
 
+#include "mem/PoolingAllocator.h"
+
+#include <map>
+
+class PeerSelector;
+class PeerSelectorPingMonitor;
+
+typedef std::pair<timeval, PeerSelector *> WaitingPeerSelector;
+/// waiting PeerSelector objects, ordered by their absolute deadlines
+typedef std::multimap<timeval, PeerSelector *, std::less<timeval>, PoolingAllocator<WaitingPeerSelector> > WaitingPeerSelectors;
```

Comment:
This line has a problem on clang at least on FreeBSD and OpenBSD, both versions 8 and 9. It fails the build with

/usr/include/c++/v1/map:1647:5: error: static_assert failed due to requirement 'is_same<std::__1::pair<timeval, PeerSelector *>, std::__1::pair<const timeval, PeerSelector *> >::value' "Allocator::value_type must be same type as value_type"

which is, indeed, correct. PoolingAllocator<WaitingPeerSelector>::value_type is WaitingPeerSelector, which doesn't match the PeerSelector * type in the multimap.
Not using the PoolingAllocator avoids the problem, and TBH for a std::pair<timeval, pointer> I'm not even sure how much benefit it brings.
Why are we even duplicating the timeval member in the key and the data? It feels to me like a clever hack which is backfiring

Reply:
> is_same<std::__1::pair<timeval, PeerSelector *>, std::__1::pair<const timeval, PeerSelector *> >

I guess that reason is a bit different: the 'timeval' does not match 'const timeval'. BTW, how could this happen that the already merged PR caused build problems? Could you please post the full log link.

Reply:
> TBH for a std::pair<timeval, pointer> I'm not even sure how much benefit it brings.

It brings about the same benefit as for most other types. In most cases, the primary benefits of memory pooling are independent of the object type.

> Why are we even duplicating the timeval member in the key and the data?

FTR: We are not duplicating the information stored in the map key. The result of ping_data::deadline() calculation is only stored as the map key. Technically, we should not even call the deadline() method twice, but that (occasional) second call was added at the last [minute](https://github.com/squid-cache/squid/pull/635#discussion_r429017121), and I did not want to risk another painful review iteration by complaining about the imperfection that did not bother the primary reviewer.